app/assets/javascripts/sitewide/drag_n_drop.js.erb

```diff
@@ -0,0 +1,350 @@
+      html += '<%= I18n.t 'assets.drag_n_drop.file_label' %>:</label> ';
+      html += truncateLongString(asset.name,
+                                 <%= Constants::FILENAME_TRUNCATION_LENGTH %>);
+      html += _validateFilesSize(asset);
+      html += '</div></div>';
+      return html;
+    }
+
+    function _removeItemHandler(id) {
+      $('[data-item-id="' + id +'"]').off('click').on('click', function(e) {
+        e.preventDefault();
+        e.stopImmediatePropagation();
+        e.stopPropagation();
+        var $el = $(this);
+        var index = $el.data('item-id');
+        totalSize -= parseInt(droppedFiles[index]/1048576);
+        droppedFiles[index] = null;
+        $el.closest('.panel-step-attachment-new').remove();
+        _validateTotalSize();
+      });
+    }
+
+    function _dragNdropAssetsOff() {
+      $('body').off('drag dragstart dragend dragover dragenter dragleave drop');
+      $('.is-dragover').hide();
+    }
+
+    return Object.freeze({
+      init: init,
+      appendFilesToForm: appendFilesToForm,
+      listItems: listItems,
+      filesStatus: filesStatus
+    });
+  })();
+
+  // Module to handle file uploading in Results
+  global.DragNDropResults = (function() {
+    var droppedFiles = [];
+    var isValid = true;
+    var totalSize = 0;
+
+    function init(files) {
+      var filesPresent = droppedFiles.length;
+      for(var i = 0; i < files.length; i++) {
+        droppedFiles.push(files[i]);
+      }
+      listItems(filesPresent);
+    }
+
+    // return the status of files if they are ready to submit
+    function filesStatus() {
+      return isValid;
+    }
+
+    // loops through a list of files and display each file in a separate panel
+    function listItems(index) {
+      _dragNdropAssetsOff();
+      for(var i = index; i < droppedFiles.length; i++) {
+          $('#new-result-assets-select')
+            .after(_uploadedAseetPreview(droppedFiles[i], i))
+            .promise()
+            .done(function() {
+              _removeItemHandler(i);
+            });
+      }
+      _validateTotalSize();
+      dragNdropAssetsInit('results');
+    }
+
+    // appent the files to the form before submit
+    function _appendFilesToForm() {
+      var regex = /result\[assets_attributes\]\[[0-9]*\]\[id\]/;
+      var prevEls = $('input').filter(function() {
+        return this.name.match(regex);
+      });
+
+      var fd = new FormData();
+      var result_names = [];
+      $.each($('input[rel="results[name]"'), function() {
+        result_names.push($(this).val());
+      });
+      result_names.reverse();
+      for(var i = 0; i < droppedFiles.length; i++) {
+        var index = i + prevEls.length;
+        var file_name = 'results_files[' + index + ']';
+        fd.append(file_name, droppedFiles[i]);
+        fd.append('results_names[' + i + ']', result_names[i]);
+      }
+      droppedFiles = [];
+      isValid = true;
+      totalSize = 0;
+      _dragNdropAssetsOff();
```

Comment:
This 4 lines could be solved with destroyAll(), if it could be used fgor steps also..